src/optimizer/query_rewrite.c

```diff
@@ -3499,6 +3500,651 @@ qo_reduce_outer_joined_tables (PARSER_CONTEXT * parser, PT_NODE * spec, PT_NODE
+		}
+
+#if defined(CUBRID_DEBUG) || defined(CUBRID_DEBUG_TEST)
+	      printf ("[debug] [ child] (1/3) table: %s\n", pt_print_alias (parser, curr_fk_spec));
+#endif
+
+	      for (curr_fk_cons = sm_class_constraints (curr_fk_mop); curr_fk_cons != NULL;
+		   curr_fk_cons = curr_fk_cons->next)
+		{
+		  bool is_skip_fk_cons = false;
+
+		  if (curr_fk_cons->type != SM_CONSTRAINT_FOREIGN_KEY)
+		    {
+		      continue;	/* curr_fk_cons->next */
+		    }
+
+		  /* WS_OID or WS_REAL_OID, which one? */
+		  if (!OID_EQ (WS_REAL_OID (curr_pk_mop), &curr_fk_cons->fk_info->ref_class_oid))
+		    {
+		      continue;	/* curr_fk_cons->next */
+		    }
+
+		  if (!BTID_IS_EQUAL (&(curr_pk_cons->index_btid), &(curr_fk_cons->fk_info->ref_class_pk_btid)))
+		    {
+		      continue;	/* curr_fk_cons->next */
+		    }
+
+#if defined(CUBRID_DEBUG) || defined(CUBRID_DEBUG_TEST)
+		  printf ("[debug] [ child] (2/3) foreign key: %s\n", curr_fk_cons->name);
+#endif
+
+		  if (append_pred_list != NULL)
+		    {
+		      parser_free_tree (parser, append_pred_list);
+		      append_pred_list = NULL;
+		    }
+
+		  for (pred_point = pred_point_list; pred_point != NULL; pred_point = pred_point->next)
+		    {
+		      SM_ATTRIBUTE *pk_cons_attr = NULL, *fk_cons_attr = NULL;
+		      PT_NODE *pk_pred_attr = NULL, *fk_pred_attr = NULL;
+		      PT_NODE *arg1 = NULL, *arg2 = NULL;
+
+		      curr_pred = pred_point;
+		      CAST_POINTER_TO_NODE (curr_pred);
+
+		      assert (PT_NODE_IS_EXPR (curr_pred));
+		      assert (PT_EXPR_OP (curr_pred) == PT_EQ);
+		      assert (PT_NODE_IS_NAME (PT_EXPR_ARG1 (curr_pred)));
+		      assert (PT_NODE_IS_NAME (PT_EXPR_ARG2 (curr_pred)));
+
+		      arg1 = PT_EXPR_ARG1 (curr_pred);
+		      arg2 = PT_EXPR_ARG2 (curr_pred);
+
+		      if (PT_SPEC_ID (curr_pk_spec) == PT_NAME_SPEC_ID (arg1))
+			{
+			  pk_pred_attr = PT_EXPR_ARG1 (curr_pred);
+			}
+		      else if (PT_SPEC_ID (curr_pk_spec) == PT_NAME_SPEC_ID (arg2))
+			{
+			  pk_pred_attr = PT_EXPR_ARG2 (curr_pred);
+			}
+		      else
+			{
+			  /* Already checked before. */
+			  assert (false);
+			  goto cleanup;
+			}
+
+		      if (PT_SPEC_ID (curr_fk_spec) == PT_NAME_SPEC_ID (arg1))
+			{
+			  fk_pred_attr = PT_EXPR_ARG1 (curr_pred);
+			}
+		      else if (PT_SPEC_ID (curr_fk_spec) == PT_NAME_SPEC_ID (arg2))
+			{
+			  fk_pred_attr = PT_EXPR_ARG2 (curr_pred);
+			}
+		      else
+			{
+			  is_skip_fk_spec = true;
+			  break;
+			}
+
+		      assert (pk_pred_attr != NULL);
+		      assert (fk_pred_attr != NULL);
+
+		      for (i = 0; curr_pk_cons->attributes[i] != NULL && curr_fk_cons->attributes[i] != NULL; i++)
+			{
+			  SM_ATTRIBUTE *pk_cons_attr = curr_pk_cons->attributes[i];
+			  SM_ATTRIBUTE *fk_cons_attr = curr_fk_cons->attributes[i];
+
+			  if (intl_identifier_casecmp (pk_cons_attr->header.name, PT_NAME_ORIGINAL (pk_pred_attr)) == 0)
+			    {
+			      if (intl_identifier_casecmp (fk_cons_attr->header.name,
+							   PT_NAME_ORIGINAL (fk_pred_attr)) == 0)
```

Comment:
Does it work properly if the order of pk and fk is different?

Reply:
The column order of the primary key and foreign key must be the same. The parent table cannot be reduced if it is compared in a different order from the column order of the foreign key referencing the primary key in the predicate.

Reply:
When creating a child table, even if foreign keys are created in the order of parent_c1 and parent_c2,
the order of the parent's primary key is changed.
```
create table parent (c1 int, c2 int, primary key (c2, c1));
create table child (c1 int primary key, parent_c1 int, parent_c2 int);
alter table child add constraint foreign key (parent_c1, parent_c2) references parent (c1, c2);
```
It is not created like "fk_t34_child_parent_c1_parent_c2 ON dba.t34_child (parent_c1, parent_c2)".
```
csql> ;sc child

  ...
 <Constraints>

     PRIMARY KEY pk_t34_child_c1 ON dba.t34_child (c1)
     FOREIGN KEY fk_t34_child_parent_c1_parent_c2 ON dba.t34_child (parent_c2, parent_c1) REFERENCES dba.t34_parent ...
```
Therefore, there is no problem in comparing the column order of the primary key and the foreign key index.

Reply:
And the creation of a foreign key for a unique index is not allowed.
```
drop table if exists child, parent;
create table parent (c1 int, c2 int, unique (c2, c1));
create table child (c1 int primary key, parent_c1 int, parent_c2 int);
alter table child add constraint foreign key (parent_c1, parent_c2) references parent (c1, c2);
```
```
ERROR: The class 'dba.parent' referred by the foreign key does not have the primary key.
```